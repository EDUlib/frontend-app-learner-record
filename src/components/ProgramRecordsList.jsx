import React from 'react';

import { ChevronLeft, Info } from '@edx/paragon/icons';
import { Alert, Hyperlink } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform/config';
import { logError } from '@edx/frontend-platform/logging';
import _ from 'lodash';

import getProgramRecords from './data/service';

const ProgramRecordsList = () => {
  const [isLoaded, setIsLoaded] = React.useState(false);
  const [hasNoData, setHasNoData] = React.useState(false);
  const [records, setRecords] = React.useState([]);

  React.useEffect(() => {
    getProgramRecords().then((data) => {
      setIsLoaded(true);
      if (_.isEmpty(data)) {
        setHasNoData(true);
      } else {
        setRecords(data.enrolled_programs);
      }
    }).catch((error) => {
      const errorMessage = (`Error: Could not fetch learner record data for user: ${error.message}`);
      logError(errorMessage);
    });
  }, []);

  const renderProfile = () => {
    const { username } = getAuthenticatedUser();
    return (
      <Hyperlink
        destination={`${getConfig().LMS_BASE_URL}/u/${username}`}
        className="mt-2 mb-4 d-inline-block muted-link pl-3 pr-3"
      >
        <ChevronLeft className="mb-1" />
        <FormattedMessage
          id="records.profile.link"
          defaultMessage="Back to My Profile"
          description="Link text that redirects logged-in user to their profile page"
        />
      </Hyperlink>
    );
  };

  const renderCredentialsServiceIssueAlert = () => (
    <div tabIndex="-1">
      <Alert variant="danger">
        <Info className="text-danger-500 mr-2 mb-1" />
        <FormattedMessage
          id="records.list.error"
          defaultMessage="An error occurred attempting to retrieve your program records. Please try again later."
          description="An error message indicating there is a problem retrieving the user's program records"
        />
      </Alert>
    </div>
  );

  const renderEmpty = () => (
    <p className="pl-3 pr-3">
      <FormattedMessage
        id="records.list.empty"
        defaultMessage="No records yet. Program records are created once you have earned at least one course
        certificate in a program."
        description="A message indicating the user has no program records to display on the Learner Records page"
      />
    </p>
  );

  const renderRows = () => (
    <section id="program-records-list" className="pl-3 pr-3">
      <header>
        <h2 className="h4">
          <FormattedMessage
            id="records.subheader"
            defaultMessage="Program Records"
            description="Subheader for the Learner Records page"
          />
        </h2>
        <p>
          <FormattedMessage
            id="records.description"
            defaultMessage="A program record is created once you have earned at least one course certificate
            in a program."
            description="Description of program records for the Learner Records page"
          />
        </p>
      </header>
    </section>
  );

  const renderList = () => {
    if (isLoaded) {
      if (hasNoData) {
        return renderCredentialsServiceIssueAlert();
      }
      if (!records.length) {
        return renderEmpty();
      }
      return renderRows();
    }
    return null;
  };

  return (
    <main id="main-content" className="pt-5 pb-5 pl-4 pr-4 " tabIndex="-1">
      {renderProfile()}
      <h1 className="h3 pl-3 pr-3 mb-4">
        <FormattedMessage
          id="records.header"
          defaultMessage="My Learner Records"
          description="Header for the Learner Records page"
        />
      </h1>
      {renderList()}
    </main>
  );
};

export default ProgramRecordsList;
