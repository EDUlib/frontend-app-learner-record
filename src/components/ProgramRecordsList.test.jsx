import React from 'react';
import axios from 'axios';
import { shallow } from 'enzyme';
import MockAdapter from 'axios-mock-adapter';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';

import ProgramRecordsList from './ProgramRecordsList';
import getProgramRecords from './data/service';

jest.mock('@edx/frontend-platform/auth');
const axiosMock = new MockAdapter(axios);
getAuthenticatedUser.mockReturnValue(axios);
axiosMock.onAny().reply(200);
axios.get = jest.fn();

jest.mock('./data/service.js');

describe('ProgramRecordsList test components', () => {
  const container = shallow(<ProgramRecordsList />);

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });

  it('should have an profile component', () => {
    expect(container.find('FormattedMessage[id="records.profile.link"]').props()).toEqual({
      id: 'records.profile.link',
      defaultMessage: 'Back to My Profile',
      description: 'Link text that redirects logged-in user to their profile page',
      values: {},
    });
  });

  it('should have a `My Learner Records` header', () => {
    expect(container.find('FormattedMessage[id="records.header"]').props()).toEqual({
      id: 'records.header',
      defaultMessage: 'My Learner Records',
      description: 'Header for the Learner Records page',
      values: {},
    });
  });

  it('should have a help section', () => {
    expect(container.find('FormattedMessage[id="records.help.header"]').props()).toEqual({
      id: 'records.help.header',
      defaultMessage: 'Questions about Learner Records?',
      description: 'Header for the help section of Learner Records page',
      values: {},
    });
  });
});

describe('ProgramRecordsList tests with completed programs', () => {
  let container;
  const wrapper = shallow(<div><ProgramRecordsList /></div>);

  beforeAll(() => {
    getProgramRecords.mockResolvedValue({
      enrolled_programs: [
        {
          name: 'myProgram',
          uuid: '12345',
          partner: 'edX',
          completed: true,
          empty: false,
        },
      ],
    });
  });

  beforeEach(() => {
    container = wrapper.find('ProgramRecordsList').dive();
  });

  it('should call get API data request', () => {
    expect(getProgramRecords).toBeCalledTimes(1);
  });

  it('should contain DataTable with completed programs', () => {
    expect(container.html()).toMatchSnapshot();
  });
});

describe('ProgramRecordsList tests with partially completed data', () => {
  let container;
  const wrapper = shallow(<div><ProgramRecordsList /></div>);

  beforeAll(() => {
    getProgramRecords.mockResolvedValue({
      enrolled_programs: [
        {
          name: 'myProgram',
          uuid: '12345',
          partner: 'edX',
          completed: false,
          empty: false,
        },
      ],
    });
  });

  beforeEach(() => {
    container = wrapper.find('ProgramRecordsList').dive();
  });

  it('should contain DataTable with partially completed programs', () => {
    expect(container.html()).toMatchSnapshot();
  });
});

describe('ProgramRecordsList tests with empty data', () => {
  let container;
  const wrapper = shallow(<div><ProgramRecordsList /></div>);

  beforeAll(() => {
    getProgramRecords.mockResolvedValue({});
  });

  beforeEach(() => {
    container = wrapper.find('ProgramRecordsList').dive();
  });

  it('should not contain a DataTable', () => {
    expect(container.html()).toMatchSnapshot();
  });
});
