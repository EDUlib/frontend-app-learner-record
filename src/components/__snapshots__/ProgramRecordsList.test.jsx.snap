// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgramRecordsList test components should match the snapshot 1`] = `"<main id=\\"main-content\\" class=\\"pt-5 pb-5 pl-4 pr-4 \\" tabindex=\\"-1\\"><a href=\\"http://localhost:18000/u/undefined\\" target=\\"_self\\" class=\\"mb-4 d-inline-block muted-link pl-3 pr-3\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"mb-1\\"><path d=\\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z\\" fill=\\"currentColor\\"></path></svg><span>Back to My Profile</span></a><h1 class=\\"h3 pl-3 pr-3 mb-4\\"><span>My Learner Records</span></h1><div class=\\"pl-3 pr-3 pt-4 pb-1\\"><h3 class=\\"h5\\"><span>Questions about Learner Records?</span></h3><span>To learn more about records you can </span><a href=\\"undefined\\" target=\\"_self\\" variant=\\"muted\\" class=\\"text-primary-900\\"><span>read more in our records help area.</span></a></div></main>"`;

exports[`ProgramRecordsList tests with completed programs should contain DataTable with completed programs 1`] = `"<main id=\\"main-content\\" class=\\"pt-5 pb-5 pl-4 pr-4 \\" tabindex=\\"-1\\"><a href=\\"http://localhost:18000/u/undefined\\" target=\\"_self\\" class=\\"mb-4 d-inline-block muted-link pl-3 pr-3\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"mb-1\\"><path d=\\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z\\" fill=\\"currentColor\\"></path></svg><span>Back to My Profile</span></a><h1 class=\\"h3 pl-3 pr-3 mb-4\\"><span>My Learner Records</span></h1><section id=\\"program-records-list\\" class=\\"pl-3 pr-3 pb-3\\"><header><h2 class=\\"h4\\"><span>Program Records</span></h2><p><span>A program record is created once you have earned at least one course certificate in a program.</span></p></header><div class=\\"pgn__data-table-wrapper\\"><div class=\\"pgn__data-table-container\\"><table role=\\"table\\" class=\\"pgn__data-table is-striped\\"><thead><tr role=\\"row\\"><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span>Program Name</span></span></th><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span>School</span></span></th><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span>Status</span></span></th><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span></span></span></th></tr></thead><tbody role=\\"rowgroup\\"><tr role=\\"row\\" class=\\"pgn__data-table-row\\"><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\">myProgram</span></td><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\">edX</span></td><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\">Completed</span></td><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\"><a href=\\"http://localhost:18150/records/programs/12345/\\" target=\\"_self\\" variant=\\"muted\\"><button type=\\"button\\" class=\\"btn btn-outline-primary\\"><span>View Program Record</span></button></a></span></td></tr></tbody></table></div></div></section><div class=\\"pl-3 pr-3 pt-4 pb-1\\"><h3 class=\\"h5\\"><span>Questions about Learner Records?</span></h3><span>To learn more about records you can </span><a href=\\"undefined\\" target=\\"_self\\" variant=\\"muted\\" class=\\"text-primary-900\\"><span>read more in our records help area.</span></a></div></main>"`;

exports[`ProgramRecordsList tests with empty data should not contain a DataTable 1`] = `"<main id=\\"main-content\\" class=\\"pt-5 pb-5 pl-4 pr-4 \\" tabindex=\\"-1\\"><a href=\\"http://localhost:18000/u/undefined\\" target=\\"_self\\" class=\\"mb-4 d-inline-block muted-link pl-3 pr-3\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"mb-1\\"><path d=\\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z\\" fill=\\"currentColor\\"></path></svg><span>Back to My Profile</span></a><h1 class=\\"h3 pl-3 pr-3 mb-4\\"><span>My Learner Records</span></h1><div tabindex=\\"-1\\"><div role=\\"alert\\" class=\\"fade alert alert-danger show\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"text-danger-500 mr-2 mb-1\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z\\" fill=\\"currentColor\\"></path></svg><span>An error occurred attempting to retrieve your program records. Please try again later.</span></div></div><div class=\\"pl-3 pr-3 pt-4 pb-1\\"><h3 class=\\"h5\\"><span>Questions about Learner Records?</span></h3><span>To learn more about records you can </span><a href=\\"undefined\\" target=\\"_self\\" variant=\\"muted\\" class=\\"text-primary-900\\"><span>read more in our records help area.</span></a></div></main>"`;

exports[`ProgramRecordsList tests with partially completed data should contain DataTable with partially completed programs 1`] = `"<main id=\\"main-content\\" class=\\"pt-5 pb-5 pl-4 pr-4 \\" tabindex=\\"-1\\"><a href=\\"http://localhost:18000/u/undefined\\" target=\\"_self\\" class=\\"mb-4 d-inline-block muted-link pl-3 pr-3\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"mb-1\\"><path d=\\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z\\" fill=\\"currentColor\\"></path></svg><span>Back to My Profile</span></a><h1 class=\\"h3 pl-3 pr-3 mb-4\\"><span>My Learner Records</span></h1><section id=\\"program-records-list\\" class=\\"pl-3 pr-3 pb-3\\"><header><h2 class=\\"h4\\"><span>Program Records</span></h2><p><span>A program record is created once you have earned at least one course certificate in a program.</span></p></header><div class=\\"pgn__data-table-wrapper\\"><div class=\\"pgn__data-table-container\\"><table role=\\"table\\" class=\\"pgn__data-table is-striped\\"><thead><tr role=\\"row\\"><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span>Program Name</span></span></th><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span>School</span></span></th><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span>Status</span></span></th><th colSpan=\\"1\\" role=\\"columnheader\\"><span class=\\"d-flex align-items-center\\"><span></span></span></th></tr></thead><tbody role=\\"rowgroup\\"><tr role=\\"row\\" class=\\"pgn__data-table-row\\"><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\">myProgram</span></td><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\">edX</span></td><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\">Partially Completed</span></td><td role=\\"cell\\"><span class=\\"pgn__data-table-cell-wrap\\"><a href=\\"http://localhost:18150/records/programs/12345/\\" target=\\"_self\\" variant=\\"muted\\"><button type=\\"button\\" class=\\"btn btn-outline-primary\\"><span>View Program Record</span></button></a></span></td></tr></tbody></table></div></div></section><div class=\\"pl-3 pr-3 pt-4 pb-1\\"><h3 class=\\"h5\\"><span>Questions about Learner Records?</span></h3><span>To learn more about records you can </span><a href=\\"undefined\\" target=\\"_self\\" variant=\\"muted\\" class=\\"text-primary-900\\"><span>read more in our records help area.</span></a></div></main>"`;
